services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/tormaystarkastelu/Persistence.kt

```diff
@@ -0,0 +1,98 @@
+package fi.hel.haitaton.hanke.tormaystarkastelu
+
+import fi.hel.haitaton.hanke.HankeEntity
+import fi.hel.haitaton.hanke.getCurrentTimeUTCAsLocalTime
+import org.springframework.data.jpa.repository.JpaRepository
+import java.time.LocalDateTime
+import javax.persistence.Embedded
+import javax.persistence.Entity
+import javax.persistence.FetchType
+import javax.persistence.GeneratedValue
+import javax.persistence.GenerationType
+import javax.persistence.Id
+import javax.persistence.JoinColumn
+import javax.persistence.ManyToOne
+import javax.persistence.Table
+
+@Entity
+@Table(name = "tormaystarkastelutulos")
+class TormaystarkasteluTulosEntity {
+    @Embedded
+    var liikennehaitta: LiikennehaittaIndeksiType? = null
+    var perus: Float? = null
+    var pyoraily: Float? = null
+    var joukkoliikenne: Float? = null
+
+    var tila: TormaystarkasteluTulosTila? = null
+    var tilaChangedAt: LocalDateTime? = null
+    var createdAt: LocalDateTime? = null
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    var id: Int? = null
+
+    @ManyToOne(optional = false, fetch = FetchType.LAZY)
+    @JoinColumn(name = "hankeid")
+    var hanke: HankeEntity? = null
+
+    // ========  functions  =======
+
+    // NOTE: there is no way to validate already invalidated result;
+    // instead, must recalculate the results and thus create a new result.
+    fun invalidate() {
+        tila = TormaystarkasteluTulosTila.EI_VOIMASSA
+        tilaChangedAt = getCurrentTimeUTCAsLocalTime()
+    }
+
+    fun addToHanke(hankeEntity: HankeEntity) {
+        if (hanke != null && hanke != hankeEntity) {
+            throw IllegalStateException("This TormaystarkasteluTulosEntity is already connected with another Hanke.")
+        }
+        hankeEntity.addTormaystarkasteluTulos(this)
+    }
+
+    fun removeFromHanke() {
+        if (hanke == null)
+            return
+        hanke!!.removeTormaystarkasteluTulos(this)
+    }
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is TormaystarkasteluTulosEntity) return false
+
+        // For persisted entities, checking id match is enough
+        if (id != null && id != other.id) return false
+
+        // For non-persisted, have to compare almost everything
+        if (hanke != null && hanke != other.hanke) return false
+        if (liikennehaitta != other.liikennehaitta) return false
+        if (perus != other.perus) return false
+        if (pyoraily != other.pyoraily) return false
+        if (joukkoliikenne != other.joukkoliikenne) return false
+        if (tila != other.tila) return false
```

Comment:
I'm not totally sure if these are part of the object's identity and should be here... but maybe equals in this case is used so that these are also needed? 

It's just that if we happen to check "Tormaystarkastelu.equals( tormaysToBeInvalidated) "let's get out of here bacause we are not handling correct tormays" " before saving to database the new state, we get the result that we are actually handling another tormaystarkastelu and not the one we wanted.

Reply:
equals() is not about object's identity; it is about which information matters when two objects are to be considered equal/interchangeable. This class was a difficult case, because the future needs would dictate certain requirements, which can not be met, yet. The well-working equals is needed in certain collection operations (where the id-field might not be available).
The example check is, afaik, not needed; there are exactly 3 operations possible with these results: calculate a new one; change state later to invalid; delete old ones (when a new results is calculated). Invalidation and deletion can only be done to a result that has already been saved to database, in which case the actions on the result-objects can be pointed with the 'id' field (without using equals). The equals is needed in the Hanke-side of the relation (a collection); current implementation should use the "if (id != null && id != other.id) return false" -part of the equals() for that.